### DISCUSSION TITLE
DataStream.WriteRange requirements.

### DISCUSSION THREAD

**Discussion by Gavin-Williams** [amerkoleci/Vortice.Windows#254] [Q&A] [answered]
Trying to use the following...

`
public class TextureWriter<T> where  T : struct
{ ...
    dataStream.WriteRange<T>(data);
`

But I'm getting 'CS8377 Type T must be non-nullable.' 

Is there something unusual about this method or is there a secret package that needs to be referenced to make T work here?

**Comment by amerkoleci** [comment]
T need to be unmanaged type or struct, not class.

**Reply by Gavin-Williams** [reply]
`public class TextureWriter<T> where  T : struct'

Doesn't that provide the proper constraint? This code works in SharpDX.

**Comment by amerkoleci** [comment, marked as answer]
Make where T : unmanaged

**Reply by Gavin-Williams** [reply]
That removes the error.

**Reply by Gavin-Williams** [reply]
So unmanaged is similar  to struct, but it enforces a stricter requirement that the struct cannot have any referenced data. It must be pure value type. That makes sense.

But then, how did SharpDX accept struct as the constraint?

Is it just the case that unmanaged wasn't used or even available with SharpDX and unmanaged provides the true intention of the constraint?

**Comment by amerkoleci** [comment]
Unmanaged makes sure your struct type is fully blittable and interop capable
